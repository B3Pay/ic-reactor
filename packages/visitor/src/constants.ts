export const TAMESTAMP_KEYS = [
  "deadline",
  "timestamp",
  "createdAt",
  "updatedAt",
  "deletedAt",
  "created_at",
  "updated_at",
  "deleted_at",
]
