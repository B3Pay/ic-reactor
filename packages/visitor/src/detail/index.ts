import { IDL } from "@dfinity/candid"
import { VisitReturnDetail } from "./rets"
import { VisitArgDetail } from "./args"

import type {
  MethodDetail,
  ServiceDetail,
  BaseActor,
  FunctionName,
} from "../types"

export * from "./rets"
export * from "./args"

export class VisitDetail<A = BaseActor> extends IDL.Visitor<
  string,
  ServiceDetail<A> | MethodDetail<A>
> {
  private argsVisitor = new VisitArgDetail<A>()
  private returnsVisitor = new VisitReturnDetail<A>()

  public visitFunc(
    t: IDL.FuncClass,
    functionName: FunctionName<A>
  ): MethodDetail<A> {
    const { detail: argDetails, ...restArgs } = this.argsVisitor.visitFunc(
      t,
      functionName
    )
    const { detail: retDetails, ...restRets } = this.returnsVisitor.visitFunc(
      t,
      functionName
    )

    return {
      argDetail: argDetails,
      retDetail: retDetails,
      ...restRets,
      ...restArgs,
    }
  }

  public visitService(t: IDL.ServiceClass): ServiceDetail<A> {
    const methodFields = t._fields.reduce((acc, services) => {
      const functionName = services[0] as FunctionName<A>
      const func = services[1]

      acc[functionName] = func.accept(this, functionName) as MethodDetail<A>

      return acc
    }, {} as ServiceDetail<A>)

    return methodFields
  }
}
