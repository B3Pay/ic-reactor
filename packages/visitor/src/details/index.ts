import type {
  FieldDetailsWithChild,
  ServiceDetails,
  FieldDetails,
  MethodDetails,
  InputDetails,
  GridLayout,
} from "./types"
import { IDL } from "@dfinity/candid"
import { isQuery } from "../helper"
import { BaseActor, FunctionName } from "@ic-reactor/core/dist/types"
import { FieldType } from "../types"

/**
 * Visit the candid file and extract the details.
 * It returns the extracted service details.
 *
 */
export class VisitDetails<A = BaseActor> extends IDL.Visitor<
  string,
  MethodDetails<A> | FieldDetailsWithChild | FieldDetails | ServiceDetails<A>
> {
  public counter = 0
  private height = 0
  private savedY = new Map<string, Map<number, number>>()
  private breakpoints = [
    { name: "xl", size: 6 },
    { name: "md", size: 4 },
    { name: "xs", size: 2 },
  ]

  public visitFunc<M extends FunctionName<A>>(
    t: IDL.FuncClass,
    functionName: M
  ): MethodDetails<A> {
    const functionType = isQuery(t) ? "query" : "update"

    this.height = 5
    const fields = t.argTypes.reduce((acc, arg, index) => {
      this.height += 2
      acc[`arg${index}`] = arg.accept(
        this,
        `arg${index}`
      ) as FieldDetailsWithChild

      return acc
    }, {} as Record<`arg${number}`, FieldDetailsWithChild | FieldDetails>)

    const h = this.height + 2
    const layouts = this.breakpoints.reduce((acc, { name, size }) => {
      const w = 2
      const x = (this.counter * w) % size
      const y = this.savedY.get(name)?.get(x) || 0
      acc[name] = {
        x,
        w,
        y,
        h,
        minH: h,
        minW: 1,
      }
      this.savedY.set(
        name,
        this.savedY.get(name)?.set(x, y + h) || new Map([[x, y + h]])
      )

      return acc
    }, {} as Record<string, GridLayout>)

    return {
      layouts,
      order: this.counter++,
      category: "home",
      functionName,
      functionType,
      __label: functionName,
      __description: t.name,
      ...fields,
    }
  }

  public visitRecord(
    t: IDL.RecordClass,
    _fields: Array<[string, IDL.Type]>,
    __label: string
  ): FieldDetailsWithChild {
    const fields = _fields.reduce((acc, [key, type]) => {
      this.height += 2
      const details = type.accept(this, key) as FieldDetailsWithChild

      acc[key] = details

      return acc
    }, {} as Record<string, FieldDetailsWithChild | FieldDetails>)

    return {
      __type: "record",
      __label,
      __hidden: false,
      __description: t.name,
      ...fields,
    }
  }

  public visitVariant(
    t: IDL.VariantClass,
    _fields: Array<[string, IDL.Type]>,
    __label: string
  ): FieldDetailsWithChild {
    this.height += 2
    let saveHeight = this.height

    const fields = _fields.reduce((acc, [key, type], index) => {
      const details = type.accept(this, key) as FieldDetailsWithChild
      saveHeight = index === 0 ? this.height + 2 : saveHeight
      acc[key] = details

      return acc
    }, {} as Record<string, FieldDetailsWithChild | FieldDetails>)

    this.height = saveHeight

    return {
      __type: "variant",
      __label,
      __description: t.name,
      ...fields,
    }
  }

  public visitTuple<T extends IDL.Type[]>(
    t: IDL.TupleClass<T>,
    components: IDL.Type[],
    __label: string
  ): FieldDetailsWithChild {
    const fields = components.reduce((acc, type, index) => {
      this.height += 2
      const details = type.accept(this, `_${index}_`) as FieldDetailsWithChild

      acc[`_${index}_`] = details

      return acc
    }, {} as Record<string, FieldDetailsWithChild | FieldDetails>)

    return {
      __type: "tuple",
      __label,
      __hidden: false,
      __description: t.name,
      ...fields,
    }
  }

  private visitedRecursive: Record<string, true> = {}
  public visitRec<T>(
    t: IDL.RecClass<T>,
    ty: IDL.ConstructType<T>,
    __label: string
  ): FieldDetailsWithChild {
    const recLabel = `${__label}_${this.counter}`
    if (!this.visitedRecursive[recLabel]) {
      this.visitedRecursive[recLabel] = true

      return ty.accept(this, __label) as FieldDetailsWithChild
    }

    return {
      __type: "recursive",
      __label,
      __hidden: false,
      __description: t.name,
    }
  }

  public visitOpt<T>(
    t: IDL.OptClass<T>,
    ty: IDL.Type<T>,
    __label: string
  ): FieldDetailsWithChild {
    const saveHeight = this.height
    const details = ty.accept(this, __label) as FieldDetailsWithChild
    this.height = saveHeight
    return {
      __type: "optional",
      __checked: false,
      __label,
      __hidden: false,
      __description: t.name,
      optional: details,
    }
  }

  public visitVec<T>(
    t: IDL.VecClass<T>,
    ty: IDL.Type<T>,
    __label: string
  ): FieldDetailsWithChild {
    const saveHeight = this.height
    const details = ty.accept(this, __label) as FieldDetailsWithChild
    this.height = saveHeight
    return {
      __type: "vector",
      __label,
      __description: t.name,
      vector: details,
    }
  }

  private visiGenericType = <T>(
    t: IDL.Type<T>,
    __type: FieldType,
    __label: string
  ): InputDetails => {
    this.height += 2
    return {
      __type,
      __label,
      __description: t.name,
    }
  }

  public visitBool(t: IDL.BoolClass, __label: string): InputDetails {
    this.height -= 2
    return this.visiGenericType(t, "boolean", __label)
  }

  public visitNull(t: IDL.NullClass, __label: string): InputDetails {
    this.height -= 2
    return this.visiGenericType(t, "null", __label)
  }

  public visitType<T>(t: IDL.Type<T>, __label: string): InputDetails {
    return this.visiGenericType(t, "unknown", __label)
  }

  public visitPrincipal(t: IDL.PrincipalClass, __label: string): InputDetails {
    return this.visiGenericType(t, "principal", __label)
  }

  public visitText(t: IDL.TextClass, label: string): InputDetails {
    return this.visiGenericType(t, "text", label)
  }

  public visitNumber<T>(t: IDL.Type<T>, __label: string): InputDetails {
    return this.visiGenericType(t, "number", __label)
  }

  public visitInt = this.visitNumber
  public visitNat = this.visitNumber
  public visitFloat = this.visitNumber
  public visitFixedInt = this.visitNumber
  public visitFixedNat = this.visitNumber

  public visitService(t: IDL.ServiceClass): ServiceDetails<A> {
    const methodDetails = t._fields.reduce((acc, services) => {
      const functionName = services[0] as FunctionName<A>
      const func = services[1]

      acc[functionName] = func.accept(this, functionName) as MethodDetails<A>

      return acc
    }, {} as ServiceDetails<A>)

    return methodDetails
  }
}
