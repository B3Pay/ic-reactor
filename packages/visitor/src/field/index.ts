import { IDL } from "@dfinity/candid"
import { VisitReturn } from "./rets"
import { VisitArg } from "./args"

import type {
  BaseActor,
  FunctionName,
  MethodField,
  NormalMethodReturn,
  ServiceField,
} from "../types"

export * from "./rets"
export * from "./args"

export class VisitField<A = BaseActor> extends IDL.Visitor<
  string,
  ServiceField<A> | MethodField<A>
> {
  private argsVisitor = new VisitArg<A>()
  private returnsVisitor = new VisitReturn<A>()

  public visitFunc(
    t: IDL.FuncClass,
    functionName: FunctionName<A>
  ): MethodField<A> {
    const {
      fields: argFields,
      defaultValues: argDefaultValues,
      ...restArgs
    } = this.argsVisitor.visitFunc(t, functionName)
    const {
      fields: retFields,
      defaultValues: retDefaultValues,
      ...restRets
    } = this.returnsVisitor.visitFunc(t, functionName) as NormalMethodReturn<A>

    return {
      argField: argFields,
      retField: retFields,
      defaultValues: {
        args: argDefaultValues,
        rets: retDefaultValues,
      },
      ...restRets,
      ...restArgs,
    }
  }

  public visitService(t: IDL.ServiceClass): ServiceField<A> {
    const methodFields = t._fields.reduce((acc, services) => {
      const functionName = services[0] as FunctionName<A>
      const func = services[1]

      acc[functionName] = func.accept(this, functionName) as MethodField<A>

      return acc
    }, {} as ServiceField<A>)

    return methodFields
  }
}
