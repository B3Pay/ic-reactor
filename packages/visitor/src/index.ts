export * from "./fields"
export * from "./details"
export * from "./layouts"
export * from "./transform"
export * from "./random/args"
export * from "./random/response"

export * as helper from "./helpers"
export * as types from "./types"
