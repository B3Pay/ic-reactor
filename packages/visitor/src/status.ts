export type StatusType = keyof (typeof Status)["_type"]

export class Status {
  /// Define the _flag object with bit flags for visibility
  private static _flag = {
    Visible: 1 << 0, /// Bit flag for visible status
    Hidden: 1 << 1, /// Bit flag for hidden status
    Disabled: 1 << 2, /// Bit flag for disabled status
  }

  /// Define the _type object with bit flags for different types
  private static _type = {
    Optional: 1 << 16, /// Bit flag for optional status
    Checked: 1 << 17, /// Bit flag for checked status
    FlexRow: 1 << 18, /// Bit flag for flex row status
  }

  /// Getter for the _flag object
  public static get Flag() {
    return this._flag
  }

  /// Getter for the _type object
  public static get Type() {
    return this._type
  }

  /// Default status is visible
  public static Default = this._flag.Visible

  /// Method to get the number representing visible status with given types
  public static Visible(...types: StatusType[]): number {
    let result = this._flag.Visible
    for (const type of types) {
      result |= this._type[type]
    }
    return result
  }

  /// Method to get the number representing hidden status with given types
  public static Hidden(...types: StatusType[]): number {
    let result = this._flag.Hidden
    for (const type of types) {
      result |= this._type[type]
    }
    return result
  }

  /// Method to get the number representing disabled status with given types
  public static Disabled(...types: StatusType[]): number {
    let result = this._flag.Disabled
    for (const type of types) {
      result |= this._type[type]
    }
    return result
  }

  /// Method to check if a status is visible
  public static isVisible(status: number): boolean {
    return (status & Status.Visible()) !== 0
  }

  /// Method to check if a status is hidden
  public static isHidden(status: number): boolean {
    return (status & this.Hidden()) !== 0
  }

  /// Method to check if a status is disabled
  public static isDisabled(status: number): boolean {
    return (status & this.Disabled()) !== 0
  }

  /// Method to check if a status is checked
  public static isChecked(status: number): boolean {
    return (status & this._type.Checked) !== 0
  }

  /// Method to check if a status is optional
  public static isOptional(status: number): boolean {
    return (status & this._type.Optional) !== 0
  }

  /// Method to toggle the checked status
  public static toggleChecked(status: number): number {
    return status ^ this._type.Checked
  }

  /// Method to toggle the visibility status
  public static toggleVisibility(status: number): number {
    return status ^ Status.Visible()
  }
}
