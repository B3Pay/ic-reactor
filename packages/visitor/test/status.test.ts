import { Status } from "../src/status"
import { createReactorStore } from "@ic-reactor/core"
import { VisitDetail } from "../src"
import { _SERVICE, idlFactory } from "./candid/ledger"

describe("createReactorStore and visit", () => {
  const { extractInterface } = createReactorStore<_SERVICE>({
    canisterId: "ryjl3-tyaaa-aaaaa-aaaba-cai",
    idlFactory,
    withVisitor: true,
  })

  const visitedDetail = () => {
    const iface = extractInterface()
    const fieldsVisitor = new VisitDetail<_SERVICE>()
    return fieldsVisitor.visitService(iface)
  }

  it("should verify detail", () => {
    const detail = visitedDetail()
    expect(detail).toMatchSnapshot()
  })
})

describe("Status Loop", () => {
  it("should print same status info", () => {
    let status = "Status:\n"
    const statusEntries = Object.entries(Status.Flag)
    statusEntries.forEach(([key, value]) => {
      status += `${key}: ${value}\n`
    })
    status += "-----\nStatusType:\n"

    const statusTypeEntries = Object.entries(Status.Prop)
    statusTypeEntries.forEach(([key, value]) => {
      status += `${key}: ${value}\n`
    })
    status += "-----\n"

    statusEntries.forEach(([key1, value1]) => {
      statusTypeEntries.forEach(([key2, value2]) => {
        status += `${key1} | ${key2}: ${value1 | value2}\n`
      })
    })

    expect(status).toMatchSnapshot()
  })
})

describe("Status", () => {
  it("should correctly identify hidden status", () => {
    const status = Status.Hidden()
    expect(Status.isHidden(status)).toBe(true)
  })

  it("should correctly identify checked status", () => {
    const status = Status.Hidden("Checked")
    expect(Status.isChecked(status)).toBe(true)
  })

  it("should correctly identify visible status", () => {
    const status = Status.Visible()
    expect(Status.isVisible(status)).toBe(true)
  })

  it("should correctly identify optional status", () => {
    const status = Status.Visible("Optional")
    expect(Status.isOptional(status)).toBe(true)
  })

  it("should correctly identify hidden optional status", () => {
    const status = Status.Hidden("Optional")
    expect(Status.isHidden(status)).toBe(true)
    expect(Status.isOptional(status)).toBe(true)
  })

  it("should correctly identify hidden checked status", () => {
    const status = Status.Hidden("Checked")
    expect(Status.isChecked(status)).toBe(true)
    expect(Status.isHidden(status)).toBe(true)
  })

  it("should correctly identify visible checked status", () => {
    const status = Status.Visible("Checked")
    expect(Status.isVisible(status)).toBe(true)
    expect(Status.isChecked(status)).toBe(true)
  })

  it("should correctly identify optional checked status", () => {
    const status = Status.Visible("Optional", "Checked")
    expect(Status.isOptional(status)).toBe(true)
    expect(Status.isChecked(status)).toBe(true)
  })

  it("should correctly toggle checked status", () => {
    let status = Status.Visible("Checked")
    expect(Status.isChecked(status)).toBe(true)
    status = Status.setUncheck(status)
    expect(Status.isChecked(status)).toBe(false)
  })

  it("should correctly toggle checked status", () => {
    let status = Status.Visible()
    status = Status.setCheck(status)
    expect(Status.isChecked(status)).toBe(true)
    status = Status.setUncheck(status)
    expect(Status.isChecked(status)).toBe(false)
  })

  it("should throw an error when hiding a non-optional status", () => {
    let status = Status.Visible()
    expect(() => Status.setHide(status)).toThrow(
      "Cannot modify a non-optional status"
    )

    status = Status.Hidden()

    expect(() => Status.setShow(status)).toThrow(
      "Cannot modify a non-optional status"
    )
  })

  it("should correctly toggle visible status", () => {
    let status = Status.Visible("Optional")
    status = Status.setHide(status)
    expect(Status.isVisible(status)).toBe(false)
  })

  it("should correctly usable using switch", () => {
    let status = Status.Visible()
    if (status === Status.Visible()) {
      expect(true).toBe(true)
    } else {
      expect(false).toBe(true)
    }

    status = Status.Hidden()
    if (status === Status.Hidden()) {
      expect(true).toBe(true)
    } else {
      expect(false).toBe(true)
    }

    status = Status.Visible("Checked")
    if (Status.isChecked(status)) {
      expect(true).toBe(true)
    } else {
      expect(false).toBe(true)
    }

    status = Status.Hidden("Checked")
    if (Status.isChecked(status)) {
      expect(true).toBe(true)
    } else {
      expect(false).toBe(true)
    }

    status = Status.Visible("Checked")
    if (status === Status.Visible("Checked")) {
      expect(true).toBe(true)
    } else {
      expect(false).toBe(true)
    }

    if (Status.isVisible(status) && Status.isChecked(status)) {
      expect(true).toBe(true)
    }

    if (Status.isHidden(status)) {
      expect(false).toBe(true)
    }

    status = Status.Visible()
    if (Status.isVisible(status) && !Status.isChecked(status)) {
      expect(true).toBe(true)
    } else {
      expect(false).toBe(true)
    }

    status = Status.Hidden("Checked")
    if (Status.isChecked(status) && Status.isHidden(status)) {
      expect(true).toBe(true)
      status = Status.setCheck(status)
      expect(Status.isChecked(status)).toBe(true)
      status = Status.setUncheck(status)
      expect(Status.isChecked(status)).toBe(false)
    } else {
      expect(false).toBe(true)
    }
  })

  it("should return props", () => {
    let status = Status.Visible("Optional")
    expect(Status.allString(status)).toStrictEqual({
      flag: "Visible",
      props: ["Optional"],
    })

    status = Status.setHide(status)

    expect(Status.isHidden(status)).toBe(true)

    expect(Status.allString(status)).toStrictEqual({
      flag: "Hidden",
      props: ["Optional"],
    })

    status = Status.setShow(status)

    expect(Status.isVisible(status)).toBe(true)

    expect(Status.allString(status)).toStrictEqual({
      flag: "Visible",
      props: ["Optional"],
    })
  })

  it("should return props", () => {
    let status = Status.Visible("Optional", "Row", "Checked")
    expect(Status.allString(status)).toStrictEqual({
      flag: "Visible",
      props: ["Optional", "Checked", "Row"],
    })

    status = Status.setHide(status)

    expect(Status.isHidden(status)).toBe(true)

    expect(Status.allString(status)).toStrictEqual({
      flag: "Hidden",
      props: ["Optional", "Checked", "Row"],
    })
  })
})
