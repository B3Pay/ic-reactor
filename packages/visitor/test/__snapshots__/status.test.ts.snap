// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Status Loop should print same status info 1`] = `
"Status:
Visible: 1
Hidden: 2
Enabled: 4
Disabled: 8
-----
StatusType:
Optional: 65536
Checked: 131072
Dynamic: 262144
-----
Visible | Optional: 65537
Visible | Checked: 131073
Visible | Dynamic: 262145
Hidden | Optional: 65538
Hidden | Checked: 131074
Hidden | Dynamic: 262146
Enabled | Optional: 65540
Enabled | Checked: 131076
Enabled | Dynamic: 262148
Disabled | Optional: 65544
Disabled | Checked: 131080
Disabled | Dynamic: 262152
"
`;

exports[`createReactorStore and visit should verify detail 1`] = `
{
  "icrc1_balance_of": {
    "argDetail": {
      "arg0": {
        "label": "__arg0",
        "record": {
          "owner": {
            "label": "owner",
            "status": 1,
          },
          "subaccount": {
            "label": "subaccount",
            "optional": {
              "label": "subaccount",
              "status": 1,
              "vector": {
                "label": "subaccount",
                "status": 1,
              },
            },
            "status": 65537,
          },
        },
        "status": 65538,
      },
    },
    "functionName": "icrc1_balance_of",
    "functionType": "query",
    "label": "icrc1_balance_of",
    "retDetail": {
      "ret0": {
        "label": "__ret0",
        "status": 1,
      },
    },
    "status": 1,
  },
  "query_blocks": {
    "argDetail": {
      "arg0": {
        "label": "__arg0",
        "record": {
          "length": {
            "label": "length",
            "status": 1,
          },
          "start": {
            "label": "start",
            "status": 1,
          },
        },
        "status": 65538,
      },
    },
    "functionName": "query_blocks",
    "functionType": "query",
    "label": "query_blocks",
    "retDetail": {
      "ret0": {
        "label": "__ret0",
        "record": {
          "archived_blocks": {
            "label": "archived_blocks",
            "status": 65537,
            "vector": {
              "label": "archived_blocks",
              "record": {
                "callback": {
                  "detail": {
                    "ret0": {
                      "label": "__ret0",
                      "status": 1,
                      "variant": {
                        "Err": {
                          "label": "Err",
                          "status": 65537,
                          "variant": {
                            "BadFirstBlockIndex": {
                              "label": "BadFirstBlockIndex",
                              "record": {
                                "first_valid_index": {
                                  "label": "first_valid_index",
                                  "status": 1,
                                },
                                "requested_index": {
                                  "label": "requested_index",
                                  "status": 1,
                                },
                              },
                              "status": 65537,
                            },
                            "Other": {
                              "label": "Other",
                              "record": {
                                "error_code": {
                                  "label": "error_code",
                                  "status": 1,
                                },
                                "error_message": {
                                  "label": "error_message",
                                  "status": 1,
                                },
                              },
                              "status": 65537,
                            },
                          },
                        },
                        "Ok": {
                          "label": "Ok",
                          "record": {
                            "blocks": {
                              "label": "blocks",
                              "status": 65537,
                              "vector": {
                                "label": "blocks",
                                "record": {
                                  "parent_hash": {
                                    "label": "parent_hash",
                                    "optional": {
                                      "label": "parent_hash",
                                      "status": 65537,
                                      "vector": {
                                        "label": "parent_hash",
                                        "status": 262145,
                                      },
                                    },
                                    "status": 65538,
                                  },
                                  "timestamp": {
                                    "label": "timestamp",
                                    "record": {
                                      "timestamp_nanos": {
                                        "label": "timestamp_nanos",
                                        "status": 1,
                                      },
                                    },
                                    "status": 1,
                                  },
                                  "transaction": {
                                    "label": "transaction",
                                    "record": {
                                      "created_at_time": {
                                        "label": "created_at_time",
                                        "record": {
                                          "timestamp_nanos": {
                                            "label": "timestamp_nanos",
                                            "status": 1,
                                          },
                                        },
                                        "status": 1,
                                      },
                                      "icrc1_memo": {
                                        "label": "icrc1_memo",
                                        "optional": {
                                          "label": "icrc1_memo",
                                          "status": 65537,
                                          "vector": {
                                            "label": "icrc1_memo",
                                            "status": 262145,
                                          },
                                        },
                                        "status": 65538,
                                      },
                                      "memo": {
                                        "label": "memo",
                                        "status": 1,
                                      },
                                      "operation": {
                                        "label": "operation",
                                        "optional": {
                                          "label": "operation",
                                          "status": 1,
                                          "variant": {
                                            "Approve": {
                                              "label": "Approve",
                                              "record": {
                                                "allowance": {
                                                  "label": "allowance",
                                                  "record": {
                                                    "e8s": {
                                                      "label": "e8s",
                                                      "status": 1,
                                                    },
                                                  },
                                                  "status": 1,
                                                },
                                                "allowance_e8s": {
                                                  "label": "allowance_e8s",
                                                  "status": 1,
                                                },
                                                "expected_allowance": {
                                                  "label": "expected_allowance",
                                                  "optional": {
                                                    "label": "expected_allowance",
                                                    "record": {
                                                      "e8s": {
                                                        "label": "e8s",
                                                        "status": 1,
                                                      },
                                                    },
                                                    "status": 1,
                                                  },
                                                  "status": 65538,
                                                },
                                                "expires_at": {
                                                  "label": "expires_at",
                                                  "optional": {
                                                    "label": "expires_at",
                                                    "record": {
                                                      "timestamp_nanos": {
                                                        "label": "timestamp_nanos",
                                                        "status": 1,
                                                      },
                                                    },
                                                    "status": 1,
                                                  },
                                                  "status": 65538,
                                                },
                                                "fee": {
                                                  "label": "fee",
                                                  "record": {
                                                    "e8s": {
                                                      "label": "e8s",
                                                      "status": 1,
                                                    },
                                                  },
                                                  "status": 1,
                                                },
                                                "from": {
                                                  "label": "from",
                                                  "status": 65537,
                                                  "vector": {
                                                    "label": "from",
                                                    "status": 262145,
                                                  },
                                                },
                                                "spender": {
                                                  "label": "spender",
                                                  "status": 65537,
                                                  "vector": {
                                                    "label": "spender",
                                                    "status": 262145,
                                                  },
                                                },
                                              },
                                              "status": 65537,
                                            },
                                            "Burn": {
                                              "label": "Burn",
                                              "record": {
                                                "amount": {
                                                  "label": "amount",
                                                  "record": {
                                                    "e8s": {
                                                      "label": "e8s",
                                                      "status": 1,
                                                    },
                                                  },
                                                  "status": 1,
                                                },
                                                "from": {
                                                  "label": "from",
                                                  "status": 65537,
                                                  "vector": {
                                                    "label": "from",
                                                    "status": 262145,
                                                  },
                                                },
                                                "spender": {
                                                  "label": "spender",
                                                  "optional": {
                                                    "label": "spender",
                                                    "status": 65537,
                                                    "vector": {
                                                      "label": "spender",
                                                      "status": 262145,
                                                    },
                                                  },
                                                  "status": 65538,
                                                },
                                              },
                                              "status": 65537,
                                            },
                                            "Mint": {
                                              "label": "Mint",
                                              "record": {
                                                "amount": {
                                                  "label": "amount",
                                                  "record": {
                                                    "e8s": {
                                                      "label": "e8s",
                                                      "status": 1,
                                                    },
                                                  },
                                                  "status": 1,
                                                },
                                                "to": {
                                                  "label": "to",
                                                  "status": 65537,
                                                  "vector": {
                                                    "label": "to",
                                                    "status": 262145,
                                                  },
                                                },
                                              },
                                              "status": 65537,
                                            },
                                            "Transfer": {
                                              "label": "Transfer",
                                              "record": {
                                                "amount": {
                                                  "label": "amount",
                                                  "record": {
                                                    "e8s": {
                                                      "label": "e8s",
                                                      "status": 1,
                                                    },
                                                  },
                                                  "status": 1,
                                                },
                                                "fee": {
                                                  "label": "fee",
                                                  "record": {
                                                    "e8s": {
                                                      "label": "e8s",
                                                      "status": 1,
                                                    },
                                                  },
                                                  "status": 1,
                                                },
                                                "from": {
                                                  "label": "from",
                                                  "status": 65537,
                                                  "vector": {
                                                    "label": "from",
                                                    "status": 262145,
                                                  },
                                                },
                                                "spender": {
                                                  "label": "spender",
                                                  "optional": {
                                                    "label": "spender",
                                                    "status": 65537,
                                                    "vector": {
                                                      "label": "spender",
                                                      "status": 262145,
                                                    },
                                                  },
                                                  "status": 65538,
                                                },
                                                "to": {
                                                  "label": "to",
                                                  "status": 65537,
                                                  "vector": {
                                                    "label": "to",
                                                    "status": 262145,
                                                  },
                                                },
                                              },
                                              "status": 65537,
                                            },
                                          },
                                        },
                                        "status": 65538,
                                      },
                                    },
                                    "status": 1,
                                  },
                                },
                                "status": 262145,
                              },
                            },
                          },
                          "status": 65537,
                        },
                      },
                    },
                  },
                  "functionName": "callback",
                  "functionType": "query",
                  "label": "callback",
                  "status": 1,
                },
                "length": {
                  "label": "length",
                  "status": 1,
                },
                "start": {
                  "label": "start",
                  "status": 1,
                },
              },
              "status": 262145,
            },
          },
          "blocks": {
            "label": "blocks",
            "status": 65537,
            "vector": {
              "label": "blocks",
              "record": {
                "parent_hash": {
                  "label": "parent_hash",
                  "optional": {
                    "label": "parent_hash",
                    "status": 65537,
                    "vector": {
                      "label": "parent_hash",
                      "status": 262145,
                    },
                  },
                  "status": 65538,
                },
                "timestamp": {
                  "label": "timestamp",
                  "record": {
                    "timestamp_nanos": {
                      "label": "timestamp_nanos",
                      "status": 1,
                    },
                  },
                  "status": 1,
                },
                "transaction": {
                  "label": "transaction",
                  "record": {
                    "created_at_time": {
                      "label": "created_at_time",
                      "record": {
                        "timestamp_nanos": {
                          "label": "timestamp_nanos",
                          "status": 1,
                        },
                      },
                      "status": 1,
                    },
                    "icrc1_memo": {
                      "label": "icrc1_memo",
                      "optional": {
                        "label": "icrc1_memo",
                        "status": 65537,
                        "vector": {
                          "label": "icrc1_memo",
                          "status": 262145,
                        },
                      },
                      "status": 65538,
                    },
                    "memo": {
                      "label": "memo",
                      "status": 1,
                    },
                    "operation": {
                      "label": "operation",
                      "optional": {
                        "label": "operation",
                        "status": 1,
                        "variant": {
                          "Approve": {
                            "label": "Approve",
                            "record": {
                              "allowance": {
                                "label": "allowance",
                                "record": {
                                  "e8s": {
                                    "label": "e8s",
                                    "status": 1,
                                  },
                                },
                                "status": 1,
                              },
                              "allowance_e8s": {
                                "label": "allowance_e8s",
                                "status": 1,
                              },
                              "expected_allowance": {
                                "label": "expected_allowance",
                                "optional": {
                                  "label": "expected_allowance",
                                  "record": {
                                    "e8s": {
                                      "label": "e8s",
                                      "status": 1,
                                    },
                                  },
                                  "status": 1,
                                },
                                "status": 65538,
                              },
                              "expires_at": {
                                "label": "expires_at",
                                "optional": {
                                  "label": "expires_at",
                                  "record": {
                                    "timestamp_nanos": {
                                      "label": "timestamp_nanos",
                                      "status": 1,
                                    },
                                  },
                                  "status": 1,
                                },
                                "status": 65538,
                              },
                              "fee": {
                                "label": "fee",
                                "record": {
                                  "e8s": {
                                    "label": "e8s",
                                    "status": 1,
                                  },
                                },
                                "status": 1,
                              },
                              "from": {
                                "label": "from",
                                "status": 65537,
                                "vector": {
                                  "label": "from",
                                  "status": 262145,
                                },
                              },
                              "spender": {
                                "label": "spender",
                                "status": 65537,
                                "vector": {
                                  "label": "spender",
                                  "status": 262145,
                                },
                              },
                            },
                            "status": 65537,
                          },
                          "Burn": {
                            "label": "Burn",
                            "record": {
                              "amount": {
                                "label": "amount",
                                "record": {
                                  "e8s": {
                                    "label": "e8s",
                                    "status": 1,
                                  },
                                },
                                "status": 1,
                              },
                              "from": {
                                "label": "from",
                                "status": 65537,
                                "vector": {
                                  "label": "from",
                                  "status": 262145,
                                },
                              },
                              "spender": {
                                "label": "spender",
                                "optional": {
                                  "label": "spender",
                                  "status": 65537,
                                  "vector": {
                                    "label": "spender",
                                    "status": 262145,
                                  },
                                },
                                "status": 65538,
                              },
                            },
                            "status": 65537,
                          },
                          "Mint": {
                            "label": "Mint",
                            "record": {
                              "amount": {
                                "label": "amount",
                                "record": {
                                  "e8s": {
                                    "label": "e8s",
                                    "status": 1,
                                  },
                                },
                                "status": 1,
                              },
                              "to": {
                                "label": "to",
                                "status": 65537,
                                "vector": {
                                  "label": "to",
                                  "status": 262145,
                                },
                              },
                            },
                            "status": 65537,
                          },
                          "Transfer": {
                            "label": "Transfer",
                            "record": {
                              "amount": {
                                "label": "amount",
                                "record": {
                                  "e8s": {
                                    "label": "e8s",
                                    "status": 1,
                                  },
                                },
                                "status": 1,
                              },
                              "fee": {
                                "label": "fee",
                                "record": {
                                  "e8s": {
                                    "label": "e8s",
                                    "status": 1,
                                  },
                                },
                                "status": 1,
                              },
                              "from": {
                                "label": "from",
                                "status": 65537,
                                "vector": {
                                  "label": "from",
                                  "status": 262145,
                                },
                              },
                              "spender": {
                                "label": "spender",
                                "optional": {
                                  "label": "spender",
                                  "status": 65537,
                                  "vector": {
                                    "label": "spender",
                                    "status": 262145,
                                  },
                                },
                                "status": 65538,
                              },
                              "to": {
                                "label": "to",
                                "status": 65537,
                                "vector": {
                                  "label": "to",
                                  "status": 262145,
                                },
                              },
                            },
                            "status": 65537,
                          },
                        },
                      },
                      "status": 65538,
                    },
                  },
                  "status": 1,
                },
              },
              "status": 262145,
            },
          },
          "certificate": {
            "label": "certificate",
            "optional": {
              "label": "certificate",
              "status": 65537,
              "vector": {
                "label": "certificate",
                "status": 262145,
              },
            },
            "status": 65538,
          },
          "chain_length": {
            "label": "chain_length",
            "status": 1,
          },
          "first_block_index": {
            "label": "first_block_index",
            "status": 1,
          },
        },
        "status": 65538,
      },
    },
    "status": 1,
  },
}
`;
